XenModel/XenAPI-Extensions/Host.cs

```diff
@@ -1301,40 +1301,54 @@ public void CheckAndPlugPBDs()
             bool allPBDsReady = false;
             int timeout = 120;
             log.DebugFormat("Waiting for PBDs on host {0} to become plugged", Name());
-            
-            do
+
+            while (timeout > 0)
             {
-                if (this.enabled)  // if the Host is not yet enabled, pbd.currently_attached may not be accurate: see CA-66496.
+                if (enabled) // if the Host is not yet enabled, pbd.currently_attached may not be accurate: see CA-66496.
                 {
                     allPBDsReady = true;
-                    foreach (PBD pbd in Connection.ResolveAll(PBDs))
+                    foreach (var pbdRef in PBDs)
                     {
-                        if (!pbd.currently_attached)
+                        var pbd = Connection.Resolve(pbdRef);
+
+                        if (pbd == null || pbd.currently_attached)
+                            continue;
+
+                        if (Helpers.StockholmOrGreater(this)) //CA-350406
                         {
-                            allPBDsReady = false;
-                            break;
+                            var sr = Connection.Resolve(pbd.SR);
+                            if (sr != null && sr.is_tools_sr)

```

Comment:
For my understanding were tools SRs introduced in Stockholm?

Reply:
The opposite: they were removed in stockholm, but on the server side the tools SR is deleted only after all the hosts have been upgraded, so we shouldn't try to plug it in.